First draft of Architecture Intent template.

DonMcCaugheyUSDS opened the pull request:
Added a `platform/engineering/collab-cycle/` directory.  Added a readme file with links to the existing Collaboration Cycle docs and to the new Architecture Intent doc.

Cribbed language from the existing Design Intent meeting doc and added the template for Architecture Intent.

donmccaughey committed:
First draft of Architecture Intent template.

Added a `platform/engineering/collab-cycle/` directory.  Added a readme
file with links to the existing Collaboration Cycle docs and to the new
Architecture Intent doc.

Cribbed language from the existing Design Intent meeting doc and added
the template for Architecture Intent.

donmccaughey committed:
Added a few more things to the template.

- Testing section
- Rollback subsection

BillChapmanUSDS commented on platform/engineering/collab-cycle/architecture-intent-template.md:
```diff
@@ -0,0 +1,56 @@
+
+Your document should be brief and high-level.  One to two pages is appropriate for many changes.  Focus on the high level and link to supporting material where appropriate; this is not a detailed engineering spec.
+
+- Product description
```
I love this section, where should we have them live?

DonMcCaugheyUSDS replied:
Added a TODO to figure out a place for these docs.  Will address in the next draft.

humancompanion-usds commented on platform/engineering/collab-cycle/architecture-intent-template.md:
```diff
@@ -0,0 +1,56 @@
+The Architecture Intent meeting helps your team build a solution that meets Veteran-facing Service Platform (VSP) engineering standards and lowers the potential for launch-blocking issues later in the development cycle.
+
+## What is the purpose of Architecture Intent meeting?
+
```
I strongly agree with Bill here. That was a problem with Design Intent that we've tried to work on. Designers initially felt like they were going to be grilled, feared the process, and then either over prepared or waited to come until the work was farther along. We worked on our docs to let them know that it is a low stakes conversation rather than a rigorous critique. That's also part of the preamble that the host of the meeting reads out at the start.

timwright12 replied:
Would it ease the tension if these were more like 1:1 meetings rather than a team meeting with a panel?

DonMcCaugheyUSDS replied:
Agreed.  Added a TODO, will rework this in the next draft.

```
TODO: emphasize that this is a low stakes, collaborative conversation to help connect eng teams with the information and resources they need to build great things
```

donmccaughey committed:
Added a "When" section.

Also added a paragraph to the template section stating that's it's okay
not to have all the answers.

Typo fixes and minor formatting changes.

timwright12 commented on platform/engineering/collab-cycle/architecture-intent-template.md:
```diff
@@ -0,0 +1,79 @@
+- You're using an architecture pattern not currently found on the va.gov platform.
+- You plan to use a new technology, library or dependency.
+- You're integrating with a new system or API, inside or outside of the VA.
+- You need to coordinate your change with other teams or your change will add work or responsiblity to other teams.
```
This one would cover virtually all work on mobile, is the intent to review a lot of small items or just large more far-reaching ones? Sometimes small API or UI changes require coordination with upstream teams and services, so I could see this requirement being hit _a lot_ to where it could back up the system.

I see this also going in the other direction when any web team (who doesn't have a mobile engineer) adds a small feature that also needs to be added to their mobile version, and the flagship team is tasked with doing that work. Do we want to insert a collab cycle meeting into that?

DonMcCaugheyUSDS replied:
Great points.  I'm trying to find a sweet spot for two cases: (1) new teams or stuck teams need help, and (2) major changes.  Small to medium sized routine changes don't need to go through this.  I've rewritten this point to:
```
- You're change requires complex coordination across teams or you need support to coordinate across teams.
 ```
Definitely open to refining this further.

timwright12 commented on platform/engineering/collab-cycle/architecture-intent-template.md:
```diff
@@ -0,0 +1,79 @@
+
+## When to schedule an Architecture Intent meeting?
+
+You should schedule an Architecture Intent meeting if any of these apply:
```
Is there a size component to this? Like, would a very large, high profile feature, that only reuses current patterns need a review?

DonMcCaugheyUSDS replied:
As above, I'm thinking of two cases: (1) new teams or stuck teams needing help, and (2) major changes.  I will work to make this more explicit in the next draft.

timwright12 commented on platform/engineering/collab-cycle/architecture-intent-template.md:
```diff
@@ -0,0 +1,79 @@
+
+## Architecture Intent Template
+
+Your document should be brief and high-level.  One to two pages is appropriate for many changes.  Focus on the high level and link to supporting material where appropriate; this is _not_ a detailed engineering spec.
```
I would just go a head and constrain it to a single page. "Your document should be brief and high-level and no more than a single page."

DonMcCaugheyUSDS replied:
Yeah, agreed.  Keep it simple.  Changed.

timwright12 commented on platform/engineering/collab-cycle/architecture-intent-template.md:
```diff
@@ -0,0 +1,79 @@
+# Architecture Intent Template
```
```suggestion
# Architecture Intent Meeting
```

Seems like the template doesn't actually start until around line 37

DonMcCaugheyUSDS replied:
Yup.  Changed.

timwright12 commented on platform/engineering/collab-cycle/architecture-intent-template.md:
```diff
@@ -0,0 +1,79 @@
+
+You should write up your Architecture Intent using the template below.  The Governance Team will need at least **2 business days** to review your materials.
+
+## Architecture Intent Template
```
Might be nice if this was broken off into separate a copy/pastable(sp?) document that's linked to from here along with an example for folks to build off.

DonMcCaugheyUSDS replied:
Agreed.  Added a todo and will break this out in the next draft.

timwright12 commented on platform/engineering/collab-cycle/architecture-intent-template.md:
```diff
@@ -0,0 +1,79 @@
+
+- Product description
+    + Brief overview of motivation for the change
+    + Link to product doc or GitHub issue
```
```suggestion
    + Link to product document or GitHub issue
```

I've just learned that sometimes people unexpectedly don't know abbreviations, so I try and avoid them

DonMcCaugheyUSDS replied:
Agreed.  Changed.

timwright12 commented on platform/engineering/collab-cycle/architecture-intent-template.md:
```diff
@@ -0,0 +1,79 @@
+    + Link to product doc or GitHub issue
+- UX design description
+    + For user-facing changes
+    + Link to lo-fi prototype or wireframes
```
```suggestion
    + Link to low fidelity prototype or wireframes
```

sanjabaj2 replied:
Are we sure we want to include UX changes in this? Those are already reviewed via design intent process.

DonMcCaugheyUSDS replied:
Revised to focus on any engineering challenges in the UX and leave UX review to the Design Intent meeting.

timwright12 commented on platform/engineering/collab-cycle/architecture-intent-template.md:
```diff
@@ -0,0 +1,79 @@
+- Data storage
+    + Describe new or modified databases, tables or columns
+    + Describe indexes and constraints
+    + Identify PII and PHI
```
```suggestion
    + Identify Personally Identifiable Information (PII) and Protected Health Information (PHI)
```

DonMcCaugheyUSDS replied:
Yup.  (Added acronym expansion at point of first use in the "When" section.)

timwright12 commented on platform/engineering/collab-cycle/architecture-intent-template.md:
```diff
@@ -0,0 +1,79 @@
+    + For user-facing changes
+    + Link to lo-fi prototype or wireframes
+- Frontend changes
+    + Identify any significant code changes
```
```suggestion
    + Identify any significant code changes
    + Identify any new design system components needed or changes to current components
```

DonMcCaugheyUSDS replied:
Yes.

timwright12 commented on platform/engineering/collab-cycle/architecture-intent-template.md:
```diff
@@ -0,0 +1,79 @@
+    + Describe automated testing
+    + Describe required test data and test user accounts
+    + Note any manual tests or user acceptance tests
+- Rollout plan
```
A main complaint I've heard that we have documentation scattered everywhere, so it might be nice think about the location of documentation at this point, so we can start to redirect teams to a more central location

DonMcCaugheyUSDS replied:
@timwright12 Strongly agree about the docs scattered everywhere.  Kicking that can down the road for this first draft, but thinking that markdown docs in `va.gov-teams` is probably the most practical central location.  I'll be soliciting opinions on this.

@sanjabaj2 Agree this is a lot for one page.  I want to find a balance between keeping this brief and making sure teams have thought through all the important things.  Changed "Test plan" to "Test strategy" and condensed to two bullet points.

mjknight50 commented on platform/engineering/collab-cycle/architecture-intent-template.md:
```diff
@@ -0,0 +1,79 @@
+
+The Architecture Intent meeting is less a formal presentation and more of a discussion.  It not only provides OCTO-DE and Platform with an early understanding of the product/feature your team wants to build, but is also an opportunity to collaborate and provide feedback on the intended implementation and surface any adjustments needed to meet VSP engineering standards. The focus is on making sure ...
+
+## When to schedule an Architecture Intent meeting?
```
I think it is really good to have such a process in place, but how are final decisions made on adoption of the new thing?  

Let's say my team wants to innovate at the VA and move to a new technology that would inarguably result in better code quality and therefore improve the veteran experience.  Let's say this new technology is called Typescript.  

When I bring this new idea to whomever is reviewing the change, are they familiar with the technology?  A lot people are afraid to admit that they just haven't used or aren't familiar with the new thing you are trying to do.  It therefore becomes very difficult to get buy-in from a group.  What happens is that a single person raising a concern out of ten meetings you have over six  months is enough to shoot down the idea.   

The idea of a single place where new ideas and technologies can be vetted and centralized is great, but I really worry that it is yet another roadblock in what is already a very difficult place to innovate.

donmccaughey committed:
Incorporated PR feedback.

Added TODO items for next draft.
Renamed main document to "Architecture Intent Meeting".

JonathanKamensDVA commented on platform/engineering/collab-cycle/architecture-intent-meeting.md:
```diff
@@ -0,0 +1,85 @@
+
+## What is the purpose of Architecture Intent meeting?
+
+The Architecture Intent meeting is less a formal presentation and more of a discussion.  It not only provides OCTO-DE and Platform with an early understanding of the product/feature your team wants to build, but is also an opportunity to collaborate and provide feedback on the intended implementation and surface any adjustments needed to meet VSP engineering standards. The focus is on making sure ...
```
"you're"

JonathanKamensDVA commented on platform/engineering/collab-cycle/architecture-intent-meeting.md:
```diff
@@ -0,0 +1,85 @@
+- You're using an architecture pattern not currently found on the va.gov platform.
+- You plan to use a new technology, library or dependency.
+- You're integrating with a new system or API, inside or outside of the VA.
+- You're change requires complex coordination across teams or you need support to coordinate across teams.
```
"Your"

JonathanKamensDVA commented on platform/engineering/collab-cycle/architecture-intent-meeting.md:
```diff
@@ -0,0 +1,85 @@
+- Data storage
+    + Describe new or modified databases, tables or columns
+    + Describe indexes and constraints
+    + Identify PII and PHI
```
"...and where and how it will be stored and processed"?
